File: app/src/main/java/com/githang/gradledoc/common/BaseListAdapter.java
package com.githang.gradledoc.common;

import android.content.Context;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;

import java.util.List;

/**
 * User: Geek_Soledad(msdx.android@qq.com)
 * Date: 2014-12-03
 * Time: 22:31
 */
public abstract class BaseListAdapter<T> extends BaseAdapter {
    private Context mContext;
    private List<T> mData;
    private int mLayoutId;

    public BaseListAdapter(Context context, List<T> data, int layoutId) {
        mContext = context;
        mData = data;
        mLayoutId = layoutId;
    }

    @Override
    public int getCount() {
        return mData.size();
    }

    @Override
    public T getItem(int position) {
        return mData.get(position);
    }

    @Override
    public long getItemId(int position) {
        return position;
    }

    @Override
    public View getView(int position, View convertView, ViewGroup parent) {
        if(convertView == null) {
            convertView = View.inflate(mContext, mLayoutId, null);
        }
        initItemView(position, convertView);
        return convertView;
    }

    public abstract void initItemView(int position, View itemView);
}

File: app/src/main/java/com/githang/gradledoc/common/recycler/BaseListAdapter.kt
package com.githang.gradledoc.common.recycler

import android.databinding.DataBindingUtil
import android.databinding.ViewDataBinding
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.ViewGroup

/**
 * @author 黄浩杭 (msdx.android@qq.com)
 * @since 2018-12-19
 */
class BaseListAdapter<D>(
        private val layoutId: Int,
        private val brId: Int
) : RecyclerView.Adapter<BindingHolder>() {
    private val list = ArrayList<D>()

    var onItemClickListener: OnItemClickListener<D>? = null

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): BindingHolder {
        val inflater = LayoutInflater.from(parent.context)
        val binding: ViewDataBinding = DataBindingUtil.inflate(inflater, layoutId, parent, false)
        val holder = BindingHolder(binding)
        return holder.apply {
            itemView.setOnClickListener {
                val position = adapterPosition
                onItemClickListener?.onItemClick(it, position, list[position])
            }
        }
    }

    override fun getItemCount(): Int {
        return list.size
    }

    override fun onBindViewHolder(holder: BindingHolder, position: Int) {
        holder.binding.setVariable(brId, list[position])
        holder.binding.executePendingBindings()
    }

    fun update(list: List<D>) {
        this.list.clear()
        this.list.addAll(list)
        notifyDataSetChanged()
    }
}